src/libraries/Common/src/System/Net/Http/Http3/QPack/QPackEncoder.cs

```diff
@@ -0,0 +1,424 @@
+        {
+            if (buffer.Length != 0)
+            {
+                buffer[0] = 0;
+                if (IntegerEncoder.Encode(s.Length, 7, buffer, out int nameLength))
+                {
+                    buffer = buffer.Slice(nameLength);
+                    if (buffer.Length >= s.Length)
+                    {
+                        EncodeValueStringPart(s, buffer);
+
+                        length = nameLength + s.Length;
+                        return true;
+                    }
+                }
+            }
+
+            length = 0;
+            return false;
+        }
+
+        /// <summary>
+        /// Encodes a value by concatenating a collection of strings, separated by a separator string.
+        /// </summary>
+        public static bool EncodeValueString(ReadOnlySpan<string> values, string separator, Span<byte> buffer, out int length)
+        {
+            if (values.Length == 1)
+            {
+                return EncodeValueString(values[0], buffer, out length);
+            }
+
+            if (values.Length == 0)
+            {
+                return EncodeValueString(string.Empty, buffer, out length);
+            }
+
+            if (buffer.Length > 0)
+            {
+                int valueLength = separator.Length * (values.Length - 1);
+                for (int i = 0; i < values.Length; ++i)
+                {
+                    valueLength += values[i].Length;
+                }
+
+                buffer[0] = 0;
+                if (IntegerEncoder.Encode(valueLength, 7, buffer, out int nameLength))
+                {
+                    buffer = buffer.Slice(nameLength);
+                    if (buffer.Length >= valueLength)
+                    {
+                        string value = values[0];
+                        EncodeValueStringPart(value, buffer);
+                        buffer = buffer.Slice(value.Length);
+
+                        for (int i = 1; i < values.Length; ++i)
+                        {
+                            EncodeValueStringPart(separator, buffer);
+                            buffer = buffer.Slice(separator.Length);
+
+                            value = values[i];
+                            EncodeValueStringPart(value, buffer);
+                            buffer = buffer.Slice(value.Length);
+                        }
+
+                        length = nameLength + valueLength;
+                        return true;
+                    }
+                }
+            }
+
+            length = 0;
+            return false;
+        }
+
+        private static void EncodeValueStringPart(string s, Span<byte> buffer)
+        {
+            Debug.Assert(buffer.Length >= s.Length);
+
+            for (int i = 0; i < s.Length; ++i)
+            {
+                char ch = s[i];
+
+                if (ch > 127)
+                {
+                    throw new QPackEncodingException("ASCII header value.");
```

Comment:
Does this string bubble out?  Let's make sure any such strings are put into the resx.